Refactor: templates

matejvasek opened the pull request:
Make `fn.Template` an interface instead of a concrete struct.

matejvasek commented:
@zroubalik @lkingland @salaboy PTAL

lkingland commented on template.go:
```diff
@@ -1,20 +1,18 @@
+	Write(ctx context.Context, f *Function) error
+}
 
+type templateConfig struct {
```
I think the concept of a `templateConfig` might be complicating what could really just be one object.

If the reason for separating these fields out is such that their fields can be embedded in Repository and Runtime structs, then I have a whole other long-winded reason why that may not be necessary (tl;dr they share fields, but are not the same because there are indeed small diferences at each level).

This might actualy be simpler and easer kept as a single `template` struct, with a few duplicated fields as necessary.

matejvasek replied:
>then I have a whole other long-winded reason why that may not be necessary (tl;dr they share fields, but are not the same because there are indeed small diferences at each level).

@lkingland what are those differences?

matejvasek replied:
To me it seems that `BuildConfig`, `HealthEndpoints`, `BuildEnvs` and `Invocation` are just defaults for `fn.Function` that can be defined at level of repository runtime, or template.
Is there any objective problem with that?

matejvasek replied:
>If the reason for separating these fields out is such that their fields can be embedded in Repository and Runtime structs,

That's exactly what I want to do in subsequent PR.

lkingland commented on template.go:
```diff
@@ -32,9 +30,76 @@ type Template struct {
 }
 
+// template
+type template struct {
```
If you do take my suggestion and keep `template` only, removing `templateConfig`, you would just need to capitalize those fields you want serialized to disk (BuildConfig, Invocation)

matejvasek replied:
see [comment](https://github.com/knative-sandbox/kn-plugin-func/pull/961#discussion_r849818376)

lkingland reviewed (approved):
I really do like this update very much.  I hope my long comments below do not obscure that fact.  The improved clarity of copying and ability to use different implementations of a Template is really going to be nice to have (especially Dynamic Templates!)

My suggestions boil down to:
* Pass data structures by value (unless sharing a memory location is actually needed)
* (related) Use non-pointer receivers for methods (unless sharing a memory location is actually needed)
* (related) Use non-pointer arguments and return new data structures with mutations applied rather than ask a method to go mutate a memory location
* Rely on `Template.Write` to decide whether or not to mask manifest.yaml, rather than masking it from the filesystem.
* Defaults should be applied in constructors (where used), with everywhere else failing fast
* Tests should use the API where possible and avoid hard-coding a specific cases or relying on a side-effect (where possible)
* Directly serialize Repository, Runtime and Template out at different levels rathern than using a shared nested config (high field overlap, but they are not the same!).
* Comments are a pain, but should really be expected, and follow the Go style of "all exported members have at least a sentence, beginning with the exported token".

/lgtm
/hold for others and for potentially update which address these above gentle suggestions

Overall looks good!

matejvasek commented on repository.go:
```diff
@@ -289,10 +289,14 @@ func repositoryRuntimes(r Repository) (runtimes []Runtime, err error) {
 func runtimeTemplates(r Repository, runtime Runtime) (templates []Template, err error) {
-	templates = []Template{}
+
+	tp := r.TemplatesPath
+	if tp == "" {
+		tp = DefaultTemplatesPath
+	}
```
I did this in future refactor in future WIP PR. But I can do it here.

matejvasek commented on filesystem.go:
```diff
@@ -145,70 +144,108 @@ func (o osFilesystem) Stat(name string) (fs.FileInfo, error) {
+	return o.fs.Stat(path.Join(o.root, name))
+}
+
+type maskingFS struct {
+	masked func(path string) bool
+	fs     Filesystem
+}
```
Well I really don't mind putting it into template code. To me this is internal code, and it IMHO really doesn't matter where we mask it.

The user of library only cares about exported function `template.Write()` the private field `template.fs` is of none concern to library user.

matejvasek commented on repository.go:
```diff
@@ -309,14 +313,23 @@ func runtimeTemplates(r Repository, runtime Runtime) (templates []Template, err
+					_, f := path.Split(p)
+					return f == templateManifest
+				},
+			},
```
>After all, the truth here is that a Template has a Filesysteem.

For our templates as of now this is true, but it won't be necessary true for future dynamic templates.
But I did move the masking to `Write()`.

matejvasek commented on client.go:
```diff
@@ -546,7 +546,7 @@ func (c *Client) Create(cfg Function) (err error) {
 	// (default builders, etc), so a new (potentially mutated) Function is
 	// returned from Templates.Write
-	f, err = c.Templates().Write(f)
+	err = c.Templates().Write(&f)
```
see [comment](https://github.com/knative-sandbox/kn-plugin-func/pull/961#issuecomment-1098383334)

matejvasek commented on templates.go:
```diff
@@ -95,76 +93,18 @@ func (t *Templates) Get(runtime, fullname string) (Template, error) {
 // of the template (which can define default Function fields, builders,
 // buildpacks, etc)
-func (t *Templates) Write(f Function) (Function, error) {
+func (t *Templates) Write(f *Function) error {
```
see [comment](https://github.com/knative-sandbox/kn-plugin-func/pull/961#issuecomment-1098383334)

matejvasek commented on template.go:
```diff
@@ -32,9 +30,76 @@ type Template struct {
+	return t.repository + "/" + t.name
+}
+
+func (t *template) Write(ctx context.Context, f *Function) error {
```
@lkingland Little bit out of topic: To be honest I have mixed feelings from `append`.
It doesn't modify slice you are appending to which is good.
However returned slice may or may not share share underlying memory.

```go
	s1 := make([]int, 1, 1024)
	s2 := append(s1, 1)

	fmt.Println(s1)
	s2[0] = 42 // it might look like you are not modifying s1 but you might be
	fmt.Println(s1)
```

It's not working like `conj` in `Clojure` where collections are immutable.

matejvasek commented:
@lkingland I addressed some of your comments:
diff: https://github.com/knative-sandbox/kn-plugin-func/compare/99df90558e9412350b0c2c79095320e6915419b4...07c0aa4566085b12372d38df621d8b2471556291

matejvasek commented:
@lkingland please re-LGMT

lkingland commented:
/lgtm

matejvasek commented:
@zroubalik @salaboy if you don't have objections I'll merge this.

matejvasek commented:
/unhold
